### PR TITLE
Add proper error for attempted use of undef slot

### PR DESCRIPTION
**Pull request by Keno**
Fixes the segfault in #50518 and turns it into a proper error at both the syntax level (to catch lowering generating bad slot references) as well as at the codegen level (to catch e.g. bad generated functions and opaque closures). However, note that the latter case is technically undefined behavior, because we do not model the possibility that an otherwise-defined argument could throw at access time. Of course, throwing an error is allowable as undefined behavior and preferable to a segfault.

### CODE REVIEW

**Comment by Keno** [comment]
https://github.com/JuliaLang/julia/pull/50559 needs to be merged first for this not to error, since we actually had a test that triggered the bad lowering.

File: src/julia-syntax.scm [hunk 1]
```diff
@@ -296,7 +296,7 @@
                                   (if (eq? n '|#self#|) (gensy) n))
                                 arg-names))))
     (let ((body (insert-after-meta body  ;; don't specialize on generator arguments
-                                   `((meta nospecialize ,@arg-names)))))
+                                   `((meta nospecialize ,@(map (lambda (idx) `(slot ,(+ idx 1))) (iota (length arg-names))))))))
```
Comment (vtjnash) [inline src/julia-syntax.scm:299]: IIRC, there was previously a bug (and open issue) that vararg arguments didn't get nospecialize'd for macros. Any idea if this fixed it?

**Review by vtjnash — APPROVED** [review]
LGTM. Though I would much rather it not be UB to access any arguments (since we might like to look at their types and type parameters to improve optimization / inlining, and it seems really awkward if that is illegal as a special case)

File: src/julia-syntax.scm (hunk 1 above)
Comment (Keno) [reply src/julia-syntax.scm:299]: I don't think so. This is for generated functions.

**Comment by Keno** [comment]
> Though I would much rather it not be UB to access any arguments

I agree with this, but that's a pre-existing issue. I wouldn't really have a problem with ripping out the whole special `#undef#` codegen handling completely, but until then, this at least makes is slightly safer.

**Pull request merged**